keps/sig-instrumentation/20200415-cardinality-enforcement.md

```diff
@@ -0,0 +1,272 @@
+---
+title: Dynamic Cardinality Enforcement
+authors:
+  - "@logicalhan"
+  - "@lilic"
+owning-sig: sig-instrumentation
+participating-sigs:
+  - sig-instrumentation
+reviewers:
+  - sig-instrumentation
+approvers:
+  - sig-instrumentation
+editor: todo
+creation-date: 2020-04-15
+last-updated: 2020-05-19
+status: provisional
+---
+
+# Dynamic Cardinality Enforcement
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Open-Question](#open-question)
+- [Graduation Criteria](#graduation-criteria)
+- [Post-Beta tasks](#post-beta-tasks)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+Background for this KEP is that metrics with unbounded dimensions can cause memory issues in the components they instrument. Our proposed solution for this is we want to be able to *dynamically configure* an allowlist of label values for a metric *at runtime*.
+
+## Motivation
+
+Having metrics turn into memory leaks is a problem, but what is even a bigger problem is  when we can't fix these issues without re-releasing the entire Kubernetes binary.
+
+Historically we have approached these issues in various ways and were not consistent. A few approaches:
+
+- Sometimes, a [metric label dimension is intended to be bound to some known sets of values but coding mistakes cause IDs to be thrown in as a label value](https://github.com/kubernetes/kubernetes/issues/53485).
+
+- In another case, [we opt to delete the entire metric](https://github.com/kubernetes/kubernetes/pull/74636) because it basically can't be used in a meaningfully correct way.
+
+- Recently we've opted to both (1) [wholesale delete a metric label](https://github.com/kubernetes/kubernetes/pull/87669) and (2) [retroactively introduce a set of acceptable values for a metric label](https://github.com/kubernetes/kubernetes/pull/87913).
+
+Fixing these issues is currently a manual process, both laborious and time-consuming. We don't have a standard prescription for resolving this class of issue.
+
+### Goals
+
+This KEP proposes a possible solution for this, with the primary goal being to enable binding a metric dimension to a known set of values in such a way that is not coupled to code releases of Kubernetes.
+
+### Non-Goals
+
+We will expose the machinery and tools to bind a metric's labels to a discrete set of values.
+
+It is *not a goal* to implement and plumb this solution for each Kubernetes component (there are many SIGs and a number of verticals, which may have their own preferred way of doing things). As such it will be up to component owners to leverage this functionality that we provide, by feeding configuration data through whatever mechanism deemed appropriate (i.e. command line flags or reading from a file).
```

Comment:
We should still track which components have adopted the mechanism and advocate for it no? If we build the mechanism and nobody ends up using it, then that kind of defeats the purpose no?

Reply:
Yes, we should track it. I am personally inclined to implement this for apimachinery, so it will be used.

Reply:
that works for me